tree/treeplayer/src/TTreeReader.cxx

```diff
@@ -222,15 +232,113 @@ void TTreeReader::Initialize()
+      }
+   }
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// Callback from TChain and TTree's LoadTree.
+
+Bool_t TTreeReader::Notify()
+{
+
+   if (fSetEntryBaseCallingLoadTree) {
+      if (fLoadTreeStatus == kExternalLoadTree) {
+         // This can happen if someone switched trees behind us.
+         // Likely cause: a TChain::LoadTree() e.g. from TTree::Process().
+         // This means that "local" should be set!
+         // There are two entities switching trees which is bad.
+         Warning("SetEntryBase()",
+                  "The current tree in the TChain %s has changed (e.g. by TTree::Process) "
+                  "even though TTreeReader::SetEntry() was called, which switched the tree "
+                  "again. Did you mean to call TTreeReader::SetLocalEntry()?",
+                  fTree->GetName());
+      }
+      fLoadTreeStatus = kInternalLoadTree;
+   } else {
+      fLoadTreeStatus = kExternalLoadTree;
+   }
+
+   if (!fEntryList && fTree->GetEntryList() && !TestBit(kBitHaveWarnedAboutEntryListAttachedToTTree)) {
+      Warning("SetEntryBase()",
+              "The TTree / TChain has an associated TEntryList. "
+              "TTreeReader ignores TEntryLists unless you construct the TTreeReader passing a TEntryList.");
+      SetBit(kBitHaveWarnedAboutEntryListAttachedToTTree);
+   }
+
+   fDirector->SetTree(fTree->GetTree());
+   if (fProxiesSet) {
+      for (auto value: fValues) {
+         value->NotifyNewTree(fTree->GetTree());
+      }
+   }
+
+   if (!fProxiesSet) {
+      return SetProxies();
+   }
+
+   return kTRUE;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// Tell readers we now have a tree.
+/// fValues gets insertions during this loop (when parameterized arrays are read),
+/// invalidating iterators. Use old-school counting instead.
+
+Bool_t TTreeReader::SetProxies() {
+
+   for (size_t i = 0; i < fValues.size(); ++i) {
+      ROOT::Internal::TTreeReaderValueBase* reader = fValues[i];
+      reader->CreateProxy();
+      if (!reader->GetProxy()){
+         return kFALSE;
+      }
+
+   }
+   // If at least one proxy was there and no error occurred, we assume the proxies to be set.
+   fProxiesSet = !fValues.empty();
+
+   // Now we need to properly set the TTreeCache. We do this in steps:
+   // 1. We set the entry range according to the entry range of the TTreeReader
+   // 2. We add to the cache the branches identifying them by the name the user provided
+   //    upon creation of the TTreeReader{Value, Array}s
+   // 3. We stop the learning phase.
+   // Operations 1, 2 and 3 need to happen in this order. See: https://sft.its.cern.ch/jira/browse/ROOT-9773?focusedCommentId=87837
+   if (fProxiesSet) {
+      const auto curFile = fTree->GetCurrentFile();
+      if (curFile && fTree->GetTree()->GetReadCache(curFile, true)) {
+         if (!(-1LL == fEndEntry && 0ULL == fBeginEntry)) {
```

Comment:
Won't this raise warnings about signed/unsigned comparisons? Maybe `~0ULL` is better than `-1LL` here.

Reply:
this is existing code ... (so no warning at the moment).